Configurable Custom Components

loicteixeira opened the pull request:
Ref #9.

Inspired from `netlify-cms`, both for the registry and for exposing React functions.

It could use a bit more documentation on how to create an entity source, but I wonder if this should be a link to `springload/draftail` documentation (which does not exists FWIW) rather that on this project.

Also, previous discussion mentioned also publishing `wagtaildraftail` as a npm package for users who have their own build step and want to go deeper.
However, the only thing that felt missing was the `ModalSource` which is a wrapper around Wagtail's jQuery chooser modals.
Sure it would also be cool to have the registry available as a normal function, but it's available globally now so not essential.
In other words, I think it would be a nice to have, not a must have.

Commit:
Add registry

Commit:
Expose Registry and basic React functions globally

Commit:
Add simple custom component to test app

Commit:
Add documentation

thibaudcolas commented on wagtaildraftail/client/wagtaildraftail.js:
```diff
@@ -54,4 +64,18 @@ const initDraftailEditor = (fieldName, options = {}) => {
+
+  // Expose basic React methods for basic needs
+  createClass: React.createClass,
+  createElement: React.createElement,
```
I don't think this is a good idea – users should be asked to use React directly to create their components (and they can do so without JSX if they wish to).

loicteixeira replied:
See my reply to the similar comment below on `tests/testapp/home/static/button_entity.js`.

thibaudcolas commented on README.rst:
```diff
@@ -17,7 +17,7 @@ Check out `Awesome Wagtail <https://github.com/springload/awesome-wagtail>`_ for
 ------------
 
-Grab the package from pip with ``pip install wagtaildraftail``, then add ``wagtaildraftail`` as an app in your Django settings.
+Grab the package from pip with ``pip install wagtaildraftail``, then add ``wagtaildraftail`` (before any other app which will try to register an entity) as an app in your Django settings.
```
This strikes me as confusing language for the most common use-case where people do not want to make their own components. Could we move this to the dedicated documentation section?

thibaudcolas commented on README.rst:
```diff
@@ -218,7 +218,79 @@ TODO
+-  an editor ``strategy`` to find the entity when the editor is loaded (implemented in JS), which is optional, the default one works fine in most cases.
+
+A ``page decorator`` is a simple Python class with a ``render`` method accepting a single argument, the ``props`` of the element. The ``render`` method can either use the ``DOM.create_element`` from ``draftjs_exporter`` (which mimic ``React.createElement``) or render some html directly. It is registered in ``DRAFT_EXPORTER_ENTITY_DECORATORS`` in the `config file <#configuration>`_.
+
```
I think I would rather have this documentation pointing at https://github.com/springload/draftjs_exporter, with the least overlap in content, so all of this content is easier to maintain over time. For this section of the documentation, this would mean not having the code sample and then for the text going along those lines:

> Decorators define how the content needs to be displayed on the site's pages, as well as within the editor.
>
> - For the pages, they are defined in Python with `draftjs_exporter`. Refer to the dedicated documentation on [the draftjs_exporter README](https://github.com/springload/draftjs_exporter#custom-components).
> - For the editor, they are defined in JS/React with `draftail` Refer to the dedicated documentation on [the Draftail README](https://github.com/springload/draftail).

We would still need to describe the `wagtaildraftail` registration / config here, since this is what this particular project "owns".

NB: I do think a big walkthrough would be valuable, but it would be more appropriate in a separate (and potentially one-off) tutorial or blog post, rather than in the README.

thibaudcolas commented on README.rst:
```diff
@@ -218,7 +218,79 @@ TODO
+
+    /* More examples at https://github.com/springload/wagtaildraftail/tree/master/wagtaildraftail/client/decorators */
+
+An ``editor source`` is usually more complex and will usually show a modal for the user to select an object or some options.
```
I'll detail this further either here or in Draftail so it's clearer what's going on.

thibaudcolas commented on tests/testapp/home/decorators.py:
```diff
@@ -0,0 +1,27 @@
+
+        anchor_properties = {
+            'className': 'button',
+            'onclick': "window.location.href='{}'".format(href),
```
I know it's just example / test code but we should follow best practice nonetheless - do not use `onclick`.

loicteixeira replied:
I was struggling to come up with an element which would reuse an existing `source` and still do something interesting, but I'll just convert this to an anchor with a regular href, explaining this is just for demo purposes.

thibaudcolas commented on tests/testapp/testapp/grains/rich_text.py:
```diff
@@ -47,6 +47,13 @@
 }
+DRAFT_ENTITY_TYPE_BUTTON = {
+    'label': 'Button',
+    'type': 'BUTTON',
```
I see this string/token is duplicated. This is the code people will likely look at / we will showcase in our examples, do we want to encourage people to set up their own constants for those types, or simply use strings? What's the most Pythonic?

loicteixeira replied:
Pythonic or not, using strings here can indeed be considered bad practice so I'll change that.

thibaudcolas commented on wagtaildraftail/client/registry.js:
```diff
@@ -0,0 +1,41 @@
+const _registry = {
```
No underscore dangle in JS.

loicteixeira replied:
This is why you shouldn't let Python developer do JS 😮

thibaudcolas commented on wagtaildraftail/client/registry.js:
```diff
@@ -0,0 +1,41 @@
+
+const getDecorator = (name) => {
+  return _registry.decorators[name];
+};
```
Should be an arrow function.

Edit: soz, `inline` arrow function.

thibaudcolas commented on wagtaildraftail/client/registry.js:
```diff
@@ -0,0 +1,41 @@
+const registerDecorator = (name, decorator) => {
+  _registry.decorators[name] = decorator;
+  console.log('decorator: ' + name);
+};
```
I think the signature of those `register` functions could be made simpler by taking advantage of object shorthands. New signature:

```js
const registerDecorators = (decorators) => {
  Object.assign(_registry.decorators, decorators);
};

window.wagtaildraftail.registerDecorators({ ButtonDecorator });
```

This has the advantage that you don't have to repeat the decorator name and its value. The other advantage is that it promotes registering all of your decorators / strategies / etc at once – we want to encourage people to separate the definition of those components from the side effect of registering them. This would look like:

```js
// In the JS entry bundle.
import ButtonDecorator from './ButtonDecorator';
import LocationDecorator from './LocationDecorator';

window.wagtaildraftail.registerDecorators({
   ButtonDecorator,
   LocationDecorator,
});
```

thibaudcolas commented on wagtaildraftail/client/registry.js:
```diff
@@ -0,0 +1,41 @@
+  getSource,
+  registerStrategy,
+  getStrategy,
+}
```
This really looks like one object rather than a bunch of named exports – could we `export default` a single object instead? Then in the respective import, `Object.assign` its methods to the `wagtailDraftail` export.

thibaudcolas commented on wagtaildraftail/client/wagtaildraftail.js:
```diff
@@ -6,11 +6,20 @@ import DraftailEditor from 'draftail';
+});
+Object.keys(sources).forEach(function(key) {
+  registerSource(key, sources[key]);
+});
```
Those two constructs would be simplified as well by using the `registerDecorators` proposal above.

thibaudcolas commented on wagtaildraftail/client/wagtaildraftail.js:
```diff
@@ -35,9 +45,9 @@ const initDraftailEditor = (fieldName, options = {}) => {
-      component: decorators[entity.decorator],
+      control: getSource(entity.source),
+      strategy: getStrategy(entity.type) || getDefaultStrategy(entity.type),
+      component: getDecorator(entity.decorator),
```
I think the `wagtaildraftail` object should be defined above, so this code can use the same `wagtaildraftail.getSource` and so on as the userland code.

thibaudcolas replied:
Additional note (that can wait for another PR): do we want to provide a way to define fallbacks for all of those things, like we do for the exporter?

loicteixeira replied:
I disagree with using `window.wagtaildraftail.getSource` in the project's code. Sure it will make it easier for a person reading the source code that it's using the exact same function that the one they use in their custom components, but it would make the whole thing less readable and very verbose for a very limited number of people (who go through the source code rather than the documentation).

No objection to fallbacks.

thibaudcolas commented on wagtaildraftail/client/wagtaildraftail.js:
```diff
@@ -54,4 +64,18 @@ const initDraftailEditor = (fieldName, options = {}) => {
 
 window.initDraftailEditor = initDraftailEditor;
 
+window.wagtaildraftail = {
```
Should this be `wagtaildraftail` or `wagtailDraftail`? The latter comes naturally to me from JS land.

thibaudcolas commented on tests/testapp/home/static/button_entity.js:
```diff
@@ -0,0 +1,6 @@
+const ButtonDecorator = ({ entityKey, children }) => {
+  const attrs = {'data-tooltip': entityKey, className: 'RichEditor-button'};
+  return window.wagtaildraftail.createElement('span', attrs, children);
+};
```
Honestly I'm not sure I want to encourage people to do this without a build step. It's fine for this PR but I might add one later on.

loicteixeira replied:
I thought we wanted to allow an easy path, but maybe that's too easy 😉

To remove this though, I'll need help with the `webpack` config, so it has 2 entry points (the admin and the test sites) and therefore 2 exit points.

thibaudcolas replied:
I think it's too brittle. To write even this small amount of JS/React without a compilation step, this means:

- Loading React on the page in a global variable.
- Writing React code without JSX.
- Writing JS code without module bundling.
- Writing JS code without transpiling.

In 2017 all of those things are bad practice IMHO. For example, this `ButtonDecorator` uses ES6 object destructuring and arrow functions so it won't work in IE11.

I'll have a think about how best to present a reusable API to create decorators and sources with the least amount of JS, for the purpose of this PR I'm happy to go ahead with this 👌

loicteixeira replied:
Agreed.

thibaudcolas reviewed (dismissed):
Thank you for taking care of this! Here is some feedback before I head to bed. I'll get back to all of this tomorrow.

Commit:
Fix import order

Commit:
Follow JS conventions for namespace

Commit:
Cleanup & Style

- Use inline arrow functions when possible.
- Remove `console.log`.
- Remove underscore dangle.

Commit:
Allow to register multiple components at once

loicteixeira commented on wagtaildraftail/client/wagtaildraftail.js:
```diff
@@ -54,4 +64,18 @@ const initDraftailEditor = (fieldName, options = {}) => {
 
 window.initDraftailEditor = initDraftailEditor;
```
@thibaudcolas: Do we want this to be `window.wagtailDraftail.initEditor` now that we have a namespace for other utils?

thibaudcolas replied:
Yup, makes sense to me!

Commit:
Export the registry as an object rather than named exports

Commit:
Encourage best practice in examples

- Do not use magic strings
- Do not use inline event handlers

Commit:
Update documentation with recent changes

Commit:
Upadte changelog

Commit:
Move editor initialisation function under wagtailDraftail namespace
